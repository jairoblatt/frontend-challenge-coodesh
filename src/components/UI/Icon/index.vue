<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
