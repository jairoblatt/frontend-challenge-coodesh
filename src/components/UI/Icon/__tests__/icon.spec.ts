import { mount, MountOptions, Wrapper } from "@vue/test-utils";
import Icon from "../index.vue";

describe("Icon unit tests", () => {
  type Instance = InstanceType<typeof Icon>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  beforeEach(() => {
    mountFunction = (config = {}) => {
      return mount(Icon, {
        ...config,
      });
    };
  });

  it("Should mount the componente", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should display element "i" when the prop icon is passed', () => {
    const wrapper = mountFunction({
      propsData: {
        icon: "mdi-eye",
      },
    });
    expect(wrapper.find("i").exists()).toBe(true);
  });

  it('Should display element "i" when the slot default is passed', () => {
    const wrapper = mountFunction({
      slots: {
        default: "mdi-eye",
      },
    });

    const iElement = wrapper.find("i");
    expect(iElement.element).toBeInstanceOf(HTMLElement);
    expect(iElement.exists()).toBe(true);
  });

  it("Should apply the css property", () => {
    const wrapper = mountFunction({
      propsData: {
        size: 20,
        color: "red",
        icon: "mdi-user",
      },
    });

    const elementStyleAttribute = wrapper.find("i").attributes().style;
    expect(elementStyleAttribute).toContain("font-size: 20px");
    expect(elementStyleAttribute).toContain("color: red");
  });
});
