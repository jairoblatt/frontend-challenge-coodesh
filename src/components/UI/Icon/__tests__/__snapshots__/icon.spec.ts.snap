// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon unit tests Should mount the componente 1`] = `""`;
